src/codal_port/microbit_microphone.c

```diff
@@ -147,13 +155,111 @@ STATIC mp_obj_t microbit_microphone_get_events(mp_obj_t self_in) {
+    enum { ARG_buffer, ARG_rate, ARG_wait, };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_buffer, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_rate, MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_wait, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = true} },
+    };
+
+    // Parse the args.
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // Check that the buffer is an AudioFrame instance.
+    if (!mp_obj_is_type(args[ARG_buffer].u_obj, &microbit_audio_frame_type)) {
+        mp_raise_TypeError(MP_ERROR_TEXT("expecting an AudioFrame"));
+    }
+    microbit_audio_frame_obj_t *audio_frame = MP_OBJ_TO_PTR(args[ARG_buffer].u_obj);
+
+    // Start recording and wait if requested.
+    microbit_microphone_record_helper(audio_frame, args[ARG_rate].u_int, args[ARG_wait].u_bool);
+
+    return mp_const_none;
+}
+static MP_DEFINE_CONST_FUN_OBJ_KW(microbit_microphone_record_into_obj, 1, microbit_microphone_record_into);
+
+static mp_obj_t microbit_microphone_is_recording(mp_obj_t self_in) {
+    (void)self_in;
+    return mp_obj_new_bool(microbit_hal_microphone_is_recording());
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(microbit_microphone_is_recording_obj, microbit_microphone_is_recording);
+
+static mp_obj_t microbit_microphone_stop_recording(mp_obj_t self_in) {
+    (void)self_in;
+    microbit_hal_microphone_stop_recording();
+    return mp_const_none;
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(microbit_microphone_stop_recording_obj, microbit_microphone_stop_recording);
+
 STATIC const mp_rom_map_elem_t microbit_microphone_locals_dict_table[] = {
+    // Methods.
+    { MP_ROM_QSTR(MP_QSTR_set_sensitivity), MP_ROM_PTR(&microbit_microphone_set_sensitivity_obj) },
     { MP_ROM_QSTR(MP_QSTR_set_threshold), MP_ROM_PTR(&microbit_microphone_set_threshold_obj) },
     { MP_ROM_QSTR(MP_QSTR_sound_level), MP_ROM_PTR(&microbit_microphone_sound_level_obj) },
     { MP_ROM_QSTR(MP_QSTR_current_event), MP_ROM_PTR(&microbit_microphone_current_event_obj) },
     { MP_ROM_QSTR(MP_QSTR_is_event), MP_ROM_PTR(&microbit_microphone_is_event_obj) },
     { MP_ROM_QSTR(MP_QSTR_was_event), MP_ROM_PTR(&microbit_microphone_was_event_obj) },
     { MP_ROM_QSTR(MP_QSTR_get_events), MP_ROM_PTR(&microbit_microphone_get_events_obj) },
+    { MP_ROM_QSTR(MP_QSTR_record), MP_ROM_PTR(&microbit_microphone_record_obj) },
+    { MP_ROM_QSTR(MP_QSTR_record_into), MP_ROM_PTR(&microbit_microphone_record_into_obj) },
+    { MP_ROM_QSTR(MP_QSTR_is_recording), MP_ROM_PTR(&microbit_microphone_is_recording_obj) },
+    { MP_ROM_QSTR(MP_QSTR_stop_recording), MP_ROM_PTR(&microbit_microphone_stop_recording_obj) },
+
+    // Constants.
+    { MP_ROM_QSTR(MP_QSTR_SENSITIVITY_LOW), MP_ROM_INT(0.079) },
+    { MP_ROM_QSTR(MP_QSTR_SENSITIVITY_MEDIUM), MP_ROM_INT(0.2) },
+    { MP_ROM_QSTR(MP_QSTR_SENSITIVITY_HIGH), MP_ROM_INT(1.0) },
```

Comment:
These should be floats:
- https://github.com/microbit-foundation/micropython-microbit-v2/issues/204

Reply:
Now fixed.